test/integration/appconfig_test.go

```diff
@@ -0,0 +1,286 @@
+								},
+							},
+						},
+					}))
+
+				return c.Create(context.Background(), ac)
+			},
+		},
+		{
+			name:   "CheckContainerizedWorkloadExists",
+			reason: "A single ContainerizedWorkload should be rendered from the ApplicationConfiguration.",
+			test: func(c client.Client) error {
+				err := waitFor(context.Background(), 3*time.Second, func() (bool, error) {
+					cw := &v1alpha2.ContainerizedWorkload{}
+					if err := c.Get(context.Background(), types.NamespacedName{Name: cwName, Namespace: defaultNS}, cw); err != nil {
+						if kerrors.IsNotFound(err) {
+							return false, nil
+						}
+						return false, err
+					}
+
+					if len(cw.Spec.Containers) != 1 {
+						return true, errors.New(errUnexpectedContainers)
+					}
+					for i, e := range cw.Spec.Containers[0].Environment {
+						if e.Name != envVars[i] {
+							return true, errors.New(errUnexpectedSubstitution)
+						}
+						if e.Value != nil && *e.Value != paramVals[i] {
+							return true, errors.New(errUnexpectedSubstitution)
+						}
+					}
+
+					return true, nil
+				})
+
+				return err
+			},
+		},
+		{
+			name:   "DeleteAppConfig",
+			reason: "An ApplicationConfiguration should be able to be deleted.",
+			test: func(c client.Client) error {
+				ac := ac(acWithName(acName), acWithNamspace(defaultNS))
+				return c.Delete(context.Background(), ac)
+			},
+		},
+		{
+			name:   "ContainerizedWorkloadShouldCleanUp",
+			reason: "A ContainerizedWorkload should be cleaned up when its ApplicationConfiguration is deleted.",
+			test: func(c client.Client) error {
+				err := waitFor(context.Background(), 3*time.Second, func() (bool, error) {
+					cw := &v1alpha2.ContainerizedWorkload{}
+					if err := c.Get(context.Background(), types.NamespacedName{Name: cwName, Namespace: defaultNS}, cw); err != nil {
+						if kerrors.IsNotFound(err) {
+							return true, nil
+						}
+						return false, err
+					}
+
+					return false, nil
+				})
+
+				return err
+			},
+		},
+	}
+
+	cfg, err := ctrl.GetConfig()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	i, err := integration.New(cfg,
+		integration.WithCRDPaths("../../crds"),
+		integration.WithCleaners(
+			integration.NewCRDCleaner(),
+			integration.NewCRDDirCleaner()),
+	)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := core.AddToScheme(i.GetScheme()); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := corev1.AddToScheme(i.GetScheme()); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := apiextensionsv1beta1.AddToScheme(i.GetScheme()); err != nil {
+		t.Fatal(err)
+	}
+
+	zl := zap.New(zap.UseDevMode(true))
+	log := logging.NewLogrLogger(zl.WithName("app-config"))
+	if err := oam.Setup(i, log); err != nil {
+		t.Fatal(err)
+	}
+
+	i.Run()
+
+	defer func() {
+		if err := i.Cleanup(); err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := tc.test(i.GetClient())
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
```

Comment:
This part seems to be pretty much the same for all tests, is there any reason to not refactoring it out?

Reply:
I think refactoring it out is a good idea. I was thinking that we would either choose the manifest pattern or building the objects in code, in which case there would only be on test file. Do you prefer one pattern to the other? Do you think there is any reason to have both?

Reply:
What is a "manifest pattern"?

Reply:
I just mean that I basically have the same test in `appconfig_test.go` and `appconfig_manifest_test.go`. The former builds the types as structs, where the second unmarshals the manifests in the `testdata` dir. I feel like it is redundant to have both. I think I prefer actually building the structs in `appconfig_test.go`

Reply:
oh, yes, I don't think we need to have both of them.

Reply:
and I like the "building the objects in the code" more if I can pick one.